packages/fast-tooling-react/src/css-property-editor/property-editor.tsx

```diff
@@ -1,237 +1,365 @@
         };
     }
 
+    public componentDidUpdate(prevProps: CSSPropertyEditorProps): void {
+        if (
+            !isEqual(this.props.data, prevProps.data) &&
+            !isEqual(this.props.data, this.editData)
+        ) {
+            // if we don't recognize data props treat as a reset
+            this.editData = isNil(this.props.data)
+                ? {}
+                : Object.assign({}, this.props.data);
+            this.setState({
+                currentEditRowUncommittedCSSPropertyKey: null,
+            });
+        }
+    }
+
+    /**
+     * Render the component
+     */
     public render(): React.ReactNode {
         return (
-            <pre
-                className={this.props.managedClasses.cssPropertyEditor}
+            <div
                 onClick={this.handleClick}
+                tabIndex={Object.keys(this.editData).length === 0 ? 0 : -1}
+                onFocus={
+                    Object.keys(this.editData).length === 0 ? this.handleEmptyFocus : null
+                }
+                className={this.props.managedClasses.cssPropertyEditor}
                 ref={this.propertyEditorRef}
             >
-                <div
+                <pre
                     className={this.props.managedClasses.cssPropertyEditor_propertyRegion}
                 >
-                    {this.renderCurrentProperties()}
-                    {this.renderNewProperty()}
-                </div>
-            </pre>
+                    {this.renderRows()}
+                </pre>
+            </div>
         );
     }
 
-    private renderCurrentProperties(): React.ReactNode {
-        if (!this.state.data) {
+    /**
+     * Render all rows
+     */
+    private renderRows(): React.ReactNode {
+        if (isNil(this.editData)) {
             return;
         }
 
-        return Object.keys(this.state.data).map(
+        return Object.keys(this.editData).map(
             (cssKey: string, index: number): React.ReactNode => {
-                return this.renderPropertyInput(cssKey, this.state.data[cssKey], index);
+                return this.renderRow(cssKey, index);
             }
         );
     }
 
-    private renderNewProperty(): React.ReactNode {
-        const newPropertyConfig: CSSPropertiesConfig = getCSSPropertyConfig({
-            [this.state.key]: this.state.value,
-        });
+    /**
+     * Render a single row
+     */
+    private renderRow = (cssKey: string, index: number): React.ReactNode => {
+        let itemKey: string = cssKey;
 
-        return this.renderNewPropertyInput(newPropertyConfig[this.state.key]);
-    }
+        if (this.currentEditRowIndex === index) {
+            if (this.currentEditRowReactKey !== null) {
+                itemKey = this.currentEditRowReactKey;
+            } else if (cssKey === "") {
+                itemKey = this.generateItemKey();
+            } else {
+                this.currentEditRowReactKey = cssKey;
+            }
+        }
```

Comment:
I would prefer to see all of this logic be in its own method, which results in the logic below just being:

```
<PropertyEditorRow
    key={this.getItemKey(key, index)}
    ... etc
/>
```

Reply:
yep